src/Parallel/GlobalCache.ci

```diff
@@ -6,18 +6,30 @@ module GlobalCache {
   include "Utilities/TaggedTuple.hpp";
   namespace Parallel {
 
+  template <typename Metavariables>
+  group[migratable] MutableGlobalCache {
+    entry MutableGlobalCache(
+        tuples::tagged_tuple_from_typelist<
+            get_mutable_global_cache_tags<Metavariables>>&);
+    template <typename GlobalCacheTag, typename Function, typename... Args>
+    entry void mutate(std::tuple<Args...> & args);
+  }
+
   template <typename Metavariables>
   nodegroup[migratable] GlobalCache {
     entry GlobalCache(
         tuples::tagged_tuple_from_typelist<
-            get_const_global_cache_tags<Metavariables>>&);
+            get_const_global_cache_tags<Metavariables>>&,
+            CProxy_MutableGlobalCache<Metavariables>&);
```

Comment:
The reason I added the proxy to the constructor was, as you say, so that you never have a GlobalCache in an ill-defined state.   

 - I don't see a disadvantage of having two constructors for testing and production versus having two functions (`set_mutable_cache_pointer` and `set_mutable_cache_proxy`) for testing and production.    
 - For the two-functions approach, I agree that there would be no need to construct a dummy mutable cache in tests that don't need a mutable cache.  This would save a line of code or so in each of those tests. But if someone later added a mutable item to those tests, a common mistake would be to forget to add a call to `set_mutable_cache_pointer`, and this would lead to runtime segfault-type errors (or hopefully runtime assertions if we added them in every member function).
 - So the only reason I would see for two-functions would be if it would make the rest of the code more consistent or more readable.

I welcome others' thoughts on this.

Reply:
I agree that the interface will cause a bit of a design disconnect to the changes needed for the load balancer changes. Unfortunately, the `set_self_proxy` can't be switched to a constructor argument. In principle, the main proxy could be sent to the constructor; although, because the tests likely won't have a good analog to that, the separate test constructor probably just wouldn't set the corresponding variable.

Would the following work?
- Declare the alternative constructor (passing in the pointer) in src, but define it in some appropriate test helper so it can't be used in production executables
- Declare the member variables as you have them
- Keep the original constructor without passing in the proxy
- set the proxy in develop with the `set_mutable_cache_proxy`, because then we can set all the proxies in a consistent way

Reply:
@moxcodes I like your proposal, especially the part about hiding the pointer-constructor in the test helpers to avoid misuse in production code.

@markscheel what are your thoughts on this?

Reply:
What @moxcodes suggests _technically_ can work, but I worry it'll lead to some of the worst bugs you can possible encounter: weak reference ODR violations. Here's what I'm worried about happening:
1. GlobalCache.hpp forward declares a constructor but doesn't define it
2. tests/.../Blah.hpp provides a definition of the constructor
3. someone forgets to include `Blah.hpp` or is unaware of it, or any other plethora of reasons why it gets missed and defines _another_ constructor definition.
4. absolute chaos with 2 weeks of debugging. Here's why: because everything is a template I believe these will all be weak references in the object files and libraries. This means that the linker is free choose _any_ definition from _any_ object file or library. It is up to us to ensure they are the same. Basically, my worry is that it is _way to easy_ to end up with an ODR violation that the linker can't actually flag and thus we end up with some of the worst UB out there.

While most of the time this won't happen, the times it does happen it will use up a very large amount of developer time.

Reply:
@nilsdeppe : interesting; I was not aware of that linker behavior -- that does sound truly nightmarish.
I suppose we'll have to define all the constructors in src, then. We can probably ask for some distinguishing feature of the metavariables so that we can include some checks that the right things are being passed in the test version vs the src executable version.

Reply:
Ah... good to know, thanks Nils!

Reply:
Added a mention of a possible way of disabling the test version from being created in `src/` to issue #2592.